File: python/efn_adapter/security.py
import base64
import hashlib
import hmac
import time


REPLAY_WINDOW = 300  # seconds


def verify_signature(api_secret: str, timestamp: str, raw_body: bytes, signature: str) -> bool:
    """Verify EFN gateway signature and timestamp freshness."""
    try:
        ts = int(timestamp)
    except (ValueError, TypeError):
        return False

    if abs(time.time() - ts) > REPLAY_WINDOW:
        return False

    message = timestamp.encode() + b"." + raw_body
    expected = "v1=" + base64.b64encode(
        hmac.new(api_secret.encode(), message, hashlib.sha256).digest()
    ).decode()
    return hmac.compare_digest(expected, signature)

File: adapters/bank-adapter/go/adapter/security.go
package adapter

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"math"
	"strconv"
	"time"
)

const replayWindow = 300 // seconds

// VerifySignature validates the EFN gateway HMAC signature and timestamp.
func VerifySignature(apiSecret, timestamp string, rawBody []byte, signature string) bool {
	ts, err := strconv.ParseInt(timestamp, 10, 64)
	if err != nil {
		return false
	}
	if math.Abs(float64(time.Now().Unix()-ts)) > replayWindow {
		return false
	}

	msg := append([]byte(timestamp+"."), rawBody...)
	mac := hmac.New(sha256.New, []byte(apiSecret))
	mac.Write(msg)
	expected := "v1=" + base64.StdEncoding.EncodeToString(mac.Sum(nil))

	return hmac.Equal([]byte(expected), []byte(signature))
}
